### PR TITLE
Move the "ConnectionInner" structure to the protocol crate

### PR DESCRIPTION
**Pull request by notgull**
This pull request does three things:

1). Move the `ConnectionInner` object from the `x11rb` crate to the new `x11rb_protocol` crate, renaming it `Connection`. This is in the aims of creating a [sans-I/O](https://sans-io.readthedocs.io/) connection.
2). Split the `WriteBuffer` object from the `ConnectionInner` type and move it into the `RustConnection` type. Output buffering should really be handled by something like `std::io::BufWriter`, although I'm not aware of one that exists for `sendmsg` (crate idea?).
3). Move some type aliases and enums into `x11rb_protocol` and then update the imports throughout.

There's certainly more protocol bits in `x11rb` that could be moved into `x11rb_protocol` at this point, and vice versa, but this is a good starting point.

### CODE REVIEW

**Review by psychon — COMMENTED** [review]
Basically :+1: , but if I approve this, I might trigger mergify.

I'd prefer if #664 is merged first and you then have a PR against the `master` branch. Otherwise, the review for #664 just gets bigger.

> There's certainly more protocol bits in x11rb that could be moved into x11rb_protocol at this point, and vice versa

From a quick look: Good further candidates are
- `x11rb::rust_connection::id_allocator`
- `x11rb::rust_connection::parse_display`

Less good candidate:
- `x11rb::rust_connection::xauth`, because it does I/O, but some sans-I/O variant of this would be great. Perhaps instead of doing file I/O, this should just have a public `get_xauthority_file_name()` function and another function that parses xauth entries from a `[u8]`. `~/.Xauthority` should be small enough that "just read everything and parse it later" should be good enough, I think.
- `x11rb::rust_connection::packet_reader` is *meant* for I/O. Perhaps something like `extra_length` function from that module is still a good idea? But that's not much code and reading connection `Setup` needs different code that would also be nice to reuse (see `read_setup()` in `x11rb::rust_connection`....
- Something based on the code in `x11rb::rust_connection::stream::DefaultStream::connect` that takes the result from `parse_display` (without refering to the `struct` explicitly) and figures out a list of "connection targets" (TCP, abstract unix socket, concrete unix socket) for connection establishment? But that would be completely new code. That would still be useful for X11 clients. Dunno what an API for that could look like.

File: x11rb-protocol/src/connection/mod.rs [hunk 1]
```diff
@@ -2,12 +2,23 @@
 
 use std::collections::VecDeque;
 
-use super::{BufWithFds, RawEventAndSeqNumber, ReplyFdKind, WriteBuffer};
-use crate::connection::{DiscardMode, SequenceNumber};
 use crate::utils::RawFdContainer;
+use crate::{DiscardMode, SequenceNumber};
+
+pub type BufWithFds = crate::BufWithFds<Vec<u8>>;
```
[the remark this answers is not in the document]
Comment (notgull) [reply x11rb-protocol/src/connection/mod.rs:8]: I agree that `Vec<u8>` is a "good enough" default for the time being.

File: x11rb-protocol/src/connection/mod.rs [hunk 2]
```diff
@@ -2,12 +2,23 @@
...
+
+/// The raw bytes of an event received by [`RustConnection`] and its sequence number.
```
[the remark this answers is not in the document]
Comment (notgull) [reply x11rb-protocol/src/connection/mod.rs:10]: Just made it a direct hardlink to `RustConnection` in the `x11rb` crate.

**Comment by notgull** [comment]
Discarding this pull request and then reopening it for `master`

**Pull request closed without merging**